server/src/test/java/org/elasticsearch/rest/RestResponseTests.java

```diff
@@ -505,6 +506,29 @@ public void testSupressedLogging() throws IOException {
         );
     }
 
+    public void testRetriableCodesShouldTakePrecedence() throws IOException {
+        RestRequest request = new FakeRestRequest();
+        RestChannel channel = new DetailedExceptionRestChannel(request);
+        ShardSearchFailure shardFailure1 = new ShardSearchFailure(
+            new ElasticsearchException("simulated"),
+            new SearchShardTarget("nodeID", new ShardId("someIndex", "someUUID", 1), null)
+        );
+
+        ShardSearchFailure shardFailure2 = new ShardSearchFailure(
+            new NodeDisconnectedException(null, "unused message", "unused action", null),
+            new SearchShardTarget("nodeID", new ShardId("someIndex", "someUUID", 2), null)
+        );
+
+        SearchPhaseExecutionException ex = new SearchPhaseExecutionException(
+            "search",
+            "all shards failed",
+            new ShardSearchFailure[] { shardFailure1, shardFailure2 }
+        );
+
+        RestResponse response = new RestResponse(channel, new RemoteTransportException("unused message", ex));
+        assertThat(response.status(), is(RestStatus.BAD_GATEWAY));
+    }
```

Comment:
could you expand to testing to cover more cases? I think we could move these to a SearchPhaseExecutionException specific test as well?